src/java/org/apache/cassandra/db/partitions/UnfilteredPartitionIterators.java

```diff
@@ -120,9 +121,16 @@ public void reduce(int idx, UnfilteredRowIterator current)
                 partitionKey = current.partitionKey();
                 isReverseOrder = current.isReverseOrder();
 
-                // Note that because the MergeListener cares about it, we want to preserve the index of the iterator.
-                // Non-present iterator will thus be set to empty in getReduced.
-                toMerge.set(idx, current);
+                if (listener != null && listener.preserveOrder())
```

Comment:
As nothing stops `preserveOrder` from changing between this call and the one in `getReduced` and `onKeyChange`, I would store the decision in a boolean and use that in the latter calls.

Better still, could we call `preserveOrder` just once at construction?

Reply:
Completed